# Collaborative Filtering Movie Recommendation


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import numpy as np
import pandas as pd


Reading in movie data from CSV

In [4]:
movies_5000 = pd.read_csv('/content/drive/MyDrive/Dataset/tmdb_5000_movies.csv')


In [5]:
movies_5000.head()


,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.sonypictures.com/movies/spectre/,206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...","[{""iso_3166_1"": ""GB"", ""name"": ""United Kingdom""...",2015-10-26,880674609,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,Spectre,6.3,4466
3,250000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 80, ""nam...",http://www.thedarkknightrises.com/,49026,"[{""id"": 849, ""name"": ""dc comics""}, {""id"": 853,...",en,The Dark Knight Rises,Following the death of District Attorney Harve...,112.312950,"[{""name"": ""Legendary Pictures"", ""id"": 923}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-07-16,1084939099,165.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,The Legend Ends,The Dark Knight Rises,7.6,9106
4,260000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://movies.disney.com/john-carter,49529,"[{""id"": 818, ""name"": ""based on novel""}, {""id"":...",en,John Carter,"John Carter is a war-weary, former military ca...",43.926995,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}]","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-03-07,284139100,132.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"Lost in our world, found in another.",John Carter,6.1,2124


In [6]:
movies_25M = pd.read_csv("/content/drive/MyDrive/Dataset/ml-25m/ml-25m/movies.csv")

In [7]:
movies_25M.head(10)


,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy
5,6,Heat (1995),Action|Crime|Thriller
6,7,Sabrina (1995),Comedy|Romance
7,8,Tom and Huck (1995),Adventure|Children
8,9,Sudden Death (1995),Action
9,10,GoldenEye (1995),Action|Adventure|Thriller


Creating a cleaner title by removing the year

In [8]:
from os import remove
import re

def clean_title(title):
   clean_title = re.sub("[^a-zA-Z0-9]", " ", title)
   remove_year = clean_title[:-5].strip()
   return remove_year

In [9]:
movies_25M["clean_title"] = movies_25M["title"].apply(clean_title)

In [10]:
movies_25M.head()

,movieId,title,genres,clean_title
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy,Toy Story
1,2,Jumanji (1995),Adventure|Children|Fantasy,Jumanji
2,3,Grumpier Old Men (1995),Comedy|Romance,Grumpier Old Men
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance,Waiting to Exhale
4,5,Father of the Bride Part II (1995),Comedy,Father of the Bride Part II


In [11]:
movies_25M.drop('title', axis = 1, inplace = True)
movies_25M.head()

,movieId,genres,clean_title
0,1,Adventure|Animation|Children|Comedy|Fantasy,Toy Story
1,2,Adventure|Children|Fantasy,Jumanji
2,3,Comedy|Romance,Grumpier Old Men
3,4,Comedy|Drama|Romance,Waiting to Exhale
4,5,Comedy,Father of the Bride Part II


In [12]:
movies_5000 = movies_5000[['title']]
movies_5000.head()

,title
0,Avatar
1,Pirates of the Caribbean: At World's End
2,Spectre
3,The Dark Knight Rises
4,John Carter


Merging the 5K dataset with the 25M dataset

In [13]:
merged_dataset = pd.merge(movies_25M, movies_5000, left_on='clean_title', right_on='title', how='inner')
merged_dataset.drop('title', axis=1, inplace=True)

merged_dataset.head()


,movieId,genres,clean_title
0,1,Adventure|Animation|Children|Comedy|Fantasy,Toy Story
1,10,Action|Adventure|Thriller,GoldenEye
2,14,Drama,Nixon
3,15,Action|Adventure|Romance,Cutthroat Island
4,16,Crime|Drama,Casino


In [14]:
len(merged_dataset)

3728

Data was lost when merging both datasets:
4809 - 3728 = 1081 (amount of data missing)

Using TFIDF vectorization technique to make each title as a numerical vector based on the words in the title

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer(ngram_range=(1,2))
tfidf = vectorizer.fit_transform(merged_dataset['clean_title'])

When search for a title, the query is converted to its own vector which is then used to match with other similar vectors by using cosine similarity. Consine similarity finds the closest vectors by cosine angle

In [16]:
from sklearn.metrics.pairwise import cosine_similarity

def search(title):
  query_vec = vectorizer.transform([title])
  similarity = cosine_similarity(query_vec, tfidf).flatten()
  indices = np.argpartition(similarity, -5)[-5:]
  results = merged_dataset.iloc[indices][::-1]
  return results

Search bar widget to find movie

In [17]:
import ipywidgets as widgets
from IPython.display import display

input = widgets.Text(
    value = "Toy Story",
    description = "Movie Title: ",
    disabled = False
)
list = widgets.Output()

def on_type(data):
  with list:
    list.clear_output()
    title = data["new"]
    if len(title) > 5:
      display(search(title))

input.observe(on_type, names='value')
display(input, list)


Text(value='Toy Story', description='Movie Title: ')

Output()

Reading in the review data from CSV

In [18]:
ratings_25M = pd.read_csv('/content/drive/MyDrive/Dataset/ml-25m/ml-25m/ratings.csv')

In [19]:
new_ratings = ratings_25M[ratings_25M['movieId'].isin(set(merged_dataset['movieId']))]

In [20]:
new_ratings.head()

,userId,movieId,rating,timestamp
0,1,296,5.0,1147880044
11,1,1653,4.0,1147868097
12,1,2011,2.5,1147868079
13,1,2012,2.5,1147868068
17,1,2573,4.0,1147878923


In [21]:
new_ratings.dtypes

,0
userId,int64
movieId,int64
rating,float64
timestamp,int64


In [22]:
movie_id = 1

Find users who liked the selected movie and recommend other movies they rated highly.
Compute a score comparing similar users vs all users and return the top recommendations.

In [23]:
def find_same_movies(movie_id):

    same_users = new_ratings[(new_ratings['movieId'] == movie_id) & (new_ratings["rating"] > 4)]["userId"].unique()

    same_user_recs = new_ratings[(new_ratings["userId"].isin(same_users)) & (new_ratings["rating"] > 4)]["movieId"]

    same_user_recs = same_user_recs.value_counts() / len(same_users)
    same_user_recs = same_user_recs[same_user_recs > .1]

    all_users = new_ratings[(new_ratings["movieId"]).isin(same_user_recs.index) & (new_ratings["rating"] > 4)]

    all_users_recs = all_users["movieId"].value_counts() / len(all_users["userId"].unique())

    rec_percentages = pd.concat([same_user_recs, all_users_recs], axis=1)
    rec_percentages.columns = ["similar", "all"]

    rec_percentages["score"] = rec_percentages["similar"] / rec_percentages["all"]

    rec_percentages = rec_percentages.sort_values("score", ascending=False)

    return rec_percentages.head(6).merge(
        merged_dataset,
        left_index=True,
        right_on="movieId"
    )[["clean_title","genres","score"]]

Widget to input movie and get recommendations

In [24]:
movie_input = widgets.Text(
    value="",
    description="Movie Title:",
    disabled=False
)

rec_list = widgets.Output()

def on_type(data):

    with rec_list:
        rec_list.clear_output()

        title = data["new"]

        if len(title) > 5:

            results = search(title)

            movie_id = results.iloc[0]["movieId"]

            display(find_same_movies(movie_id))

movie_input.observe(on_type, names="value")

display(movie_input, rec_list)

Text(value='', description='Movie Title:')

Output()